### DISCUSSION TITLE
Map composite primary key to separate data class

### DISCUSSION THREAD

**Discussion by vladimirfx** [komapper/komapper#686] [Ideas]
In our entities we use such mapping of composite primary keys:

```kotlin
data class Entity(
  val id: PK,
  val someProp: String
){
  data class PK(val k: Int, val v: String)
}
```

Can it be implemented in komapper? 
May some workaround exist?

**Comment by nakamura-to** [comment]
You can write it using a secondary constructor as follows:

```kotlin
@KomapperEntity
data class Entity(
    @KomapperId
    val id1: Int,
    @KomapperId
    val id2: String,
    val someProp: String
){
    constructor(id: PK, someProp: String): this(id.k, id.v, someProp)
    
    data class PK(val k: Int, val v: String)
}
```

**Reply by vladimirfx** [reply]
This looks like acceptable workaround.
Do you plan first class support for embeddables and composite id?

**Reply by nakamura-to** [reply]
Sorry, we have no plans to do so.
We want to avoid the support because it brings implementation complexity.

**Reply by vladimirfx** [reply]
Maybe multicolumn converter?

**Reply by nakamura-to** [reply]
There are no plans to provide a multicolumn converter too.

Another solution is splitting your entity into table and domain models:

```kotlin
@KomapperEntity
data class TableModel(
    @KomapperId
    val id1: Int,
    @KomapperId
    val id2: String,
    val name: String,
)
```

```kotlin
data class DomainModel(
    val pk: PK,
    val name: String,
) {
    data class PK(val k: Int, val v: String)
}
```

You can convert each other in your application:

```kotlin
val m = Meta.tableModel
val query = QueryDsl.from(m)
val domainModels = database.runQuery(query)
    .map { DomainModel(DomainModel.PK(it.id1, it.id2), it.name) }
```

**Reply by vladimirfx** [reply]
Certainly, I can. But in this way I can't justify usage of Komapper instead of plain R2DBC or Spring Data R2DBC. I already wrote mappers by hands and search for advanced mapping solutions with embeddable support. Komapper looks like ORM of dream with it typesafe external mapping and very efficient generated mappers. 
OK, thanks you for responses and good product. I proceed my findings.

**Reply by nakamura-to** [reply]
We have reconsidered and decided to support [Embedded Value](https://www.martinfowler.com/eaaCatalog/embeddedValue.html) in the next version. Please see the working pull request #702  if you would like.

Your sample code can be rewritten in the future as follows:

```kotlin
@KomapperEntity
data class Entity(
  @KomapperEmbeddedId
  val id: PK,
  val someProp: String
){
  data class PK(val k: Int, val v: String)
}
```

**Reply by vladimirfx** [reply]
I've look at PR - it should be enough for most cases. For rest 4% multicolumn convert will be perfect. Or ability to set converter in column override annotations.

**Reply by nakamura-to** [reply]
I've released new version.
https://github.com/komapper/komapper/releases/tag/v1.3.0

Please try it out!
